accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,191 @@
+
+While adding a new slot is a non-issue, the movement or removal of a window needs to be handled properly.
+
+As it is nearly impossible to identify a change of a special event we need to follow a different strategy.
+When we refresh or upload a changed ICalendar file, we need to re-evaluate all scheduled actions of systems using the
+new schedule.
+We also need to provide a resolve strategy. This can happen while implementing modules to move actions
+to a different time. Possible modules could be:
+
+- Nearest: Use the nearest maintenance window
+- Previous: Use the maintenance window before the scheduled date
+- Next: Use the maintenance window after the scheduled date
+- Cancel: Cancel the Action
+- Fail: do not import the new schedule at all and keep the existing one
+
+The customer can configure the order:
+
+  resolveOrder=Nearest,Previous,Next,Fail
+
+
+### Schedule an Action
+
+When a system is assigned to a schedule, only schedules withing a maintenance window should be allowed.
+In the User Interface we should show in these cases not the datepicker, but two simple Combo boxes with Date and Time.
+The Date contains the next 20 dates where a maintenance window exists. Depending on the date box the available
+times in the time box change.
+
+When using SSM it is possible that systems have different maintenance windows. The combination systems belong
+to one maintenance schedule and systems without maintenance schedule is not critical. The action would be scheduled
+at a time which is inside of a maintenance window of the systems which belong to the schedule.
+
+Systems in different schedules needs special handling:
+
+1. do not allow systems with different schedules in SSM
+2. divide systems by maintenance schedules and display a combo box with the names. Let the user choose for which set he want to schedule it. He need to repeat it for the rest.
+3. show the picker multiple times; for every maintenance schedule one.
+
+(1 limiting too much and 3 might look ugly - I would prefer Option 2)
+
+
+### Assign a Maintenance Schedule to a System
+
+There could be multiple ways to assign systems to maintenance schedules.
+
+- In the systems details page we need to show the selected schedule with a view and a change option.
+- It should be possible to change the maintenance schedule for multiple systems using SSM
+- In the details of the Maintenance Schedule we could have a systems page where add/remove is possible.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+
+## What does it mean to turn maintenance mode "on" and "off"
+
+- A flag in DB which needs to be changed?
+- We evaluate the calendar and calculate the state of the system?
+- mix: taskomatic job (every minute) which evaluate the calendar and switch the flag (and trigger any other action which is needed)
+
```

Comment:
The part that's missing here is that systems that are using Maintenance Windows should probably be locked (like cluster nodes) while they're _not_ in maintenance mode. This will likely affect the implementation, as the locking needs to be applied and unapplied (so just doing a DB query `is_in_maintenance_mode` or so won't do the job.

Reply:
Well, we put exactly that to future iterations as this brings a lot of extra work.
When doing this in the first iteration we may not be able to deliver in time.

This is also the reason why it is in the open questions section.

On the other hand I am not 100% sure if this is really needed in the end.
Without it, it is a "SUSE Manager" feature and when you connect directly via salt nobody will hold you back on doing stupid things.
But maybe this is good enough for a lot of people. I think we will find out via feedback on this.

Reply:
So do I get it correctly that this will be enforced for Actions only, and not at the Salt level?

Reply:
At least initially, yes. As written above, the "blackouting" (locking on the salt modules level) the minions **could** come up later (2nd iteration).

Reply:
Just found out that we might not be able to enforce this on minion level.
It would effect docker and kiwi image building. But these actions should happen when the system is **not** in maintenance mode.
If we really want to enforce it, we need a big refactoring and need to implement an own daemon to build images and exclude it from `salt`.

Reply:
From a tech POV, I presently cannot imagine a unique solution to enforce this both at the Action level and the Salt level - I guess we will always somehow need two almost-separate code changes, as underlying mechanisms are completely different.

From a product POV, I also recognize this concept clicks much better with the Action model, and could be seen as added value on top of plain Salt.

On the other hand though, I see users having to or preferring to avoid Actions and work directly with Salt, at least in part, in many cases.

So my only recommendation would be: if possible, try to confirm from more-user-oriented colleagues that addressing Actions first is actually more valuable than having an (initial) Salt-only solution (I can easily imagine that being actually the case, just playing devil's advocate here).